### ISSUE TITLE
"PickleBuffer" is incompatible with protocol "SupportsBytes"

### ISSUE THREAD

**Issue by jakirkham** [microsoft/pyright#4043]
**Describe the bug**

Passing a `PickleBuffer` instance, like `pb`, to the `bytes` constructor, like `bytes(pb)`, is valid and works. However this raises a typing error with pyright.

```python
error: Argument of type "PickleBuffer" cannot be assigned to parameter "__o" of type "SupportsBytes" in function "__new__"
    "PickleBuffer" is incompatible with protocol "SupportsBytes"
      "__bytes__" is not present (reportGeneralTypeIssues)
```

**To Reproduce**

Taking the script below and calling `pyright script.py`...

```python
# script.py


from array import array
from pickle import PickleBuffer


class MyArray(array):
    @classmethod
    def from_array(cls, typecode, initializer):
        if isinstance(initializer, PickleBuffer):
            initializer = bytes(initializer)
        return cls(typecode, initializer)

    def __reduce_ex__(self, protocol):
        if protocol >= 5:
            return type(self).from_array, (self.typecode, PickleBuffer(self)), None
        else:
            return type(self).from_array, (self.typecode, bytes(self)), None
```

...causes the following error...

```python
script.py:12:33 - error: Argument of type "PickleBuffer" cannot be assigned to parameter "__o" of type "SupportsBytes" in function "__new__"
    "PickleBuffer" is incompatible with protocol "SupportsBytes"
      "__bytes__" is not present (reportGeneralTypeIssues)
```

**Expected behavior**

No error would occur

**Screenshots or Code**

Already included above.

**VS Code extension or command-line**

Via command-line. Using 1.1.275. Installed from conda-forge.

**Additional context**

Edit: Raised upstream as issue ( https://github.com/python/typeshed/issues/8889 ) as well.

**Comment by erictraut** [comment]
Pyright is correct in generating an error in this case. An argument of `PickleBuffer` does not satisfy any of the overloaded constructor signatures for `bytes`.  Here are the supported signatures according to the `builtins.pyi` stub file in typeshed:

```python
    @overload
    def __new__(cls: type[Self], __ints: Iterable[SupportsIndex]) -> Self: ...
    @overload
    def __new__(cls: type[Self], __string: str, encoding: str, errors: str = ...) -> Self: ...
    @overload
    def __new__(cls: type[Self], __length: SupportsIndex) -> Self: ...
    @overload
    def __new__(cls: type[Self]) -> Self: ...
    @overload
    def __new__(cls: type[Self], __o: SupportsBytes) -> Self: ...
```

Of these five overloaded signatures, only two of them are potential matches — the third and fifth. The third signature requires an argument that implements the `SupportsIndex` protocol, and the fifth requires an argument that implements the `SupportsBytes` protocol. The `PickleBuffer` type doesn't match either of these protocols.

If you think that `PickleBuffer` should work in this case, you could file a bug report in the [typeshed](https://github.com/python/typeshed) project. Pyright (and other Python type checkers) get their core type stubs from here.

**Comment by jakirkham** [comment]
Yep, thanks Eric! At the bottom of the OP there is a link to the upstream issue. Feel free to follow along if it is of interest :)

**Comment by jakirkham** [comment]
FWIW this was fixed upstream in PR ( https://github.com/python/typeshed/pull/8924 ). Not sure if there are other things needed for pyright to use that change; so, wanted to mention this here.

**Comment by erictraut** [comment]
@jakirkham, thanks for the heads up. I integrate the latest typeshed changes every couple of weeks. I'll make sure to include this in the next pyright release.

**Comment by jakirkham** [comment]
Thanks Eric! That timeline helps. Will keep an eye out for the next update.
